CB-14398 - Add a RecoveryService to choose between Upgrade and Resize recoveries.

brycederriso opened the pull request:
Closes [CB-14398](https://jira.cloudera.com/browse/CB-14398).

Adds a RecoveryService which chooses one of Resize or Upgrade recovery.

Change the recovery controller to call into the Recovery Service.

This is small PR, additional changes are forthcoming, depending on results of design doc review.

yg-gd commented on datalake/src/main/java/com/sequenceiq/datalake/service/resize/recovery/ResizeRecoveryService.java:
```diff
@@ -0,0 +1,46 @@
+        throw new BadRequestException("SDX Resize Recovery is not yet implemented");
+    }
+
+    public SdxRecoveryResponse triggerRecovery() {
```
These methods could be defined in a generic recovery service interface. along with the same parameters as sdxUpgradeRecoveryService

brycederriso replied:
I've played around a bit with this and I'm not sure I have a diverse enough implementation to warrant extracting an interface yet. We've got 3 things that look like they have the same functionality, but one of them just delegates to one of the other two:
* `RecoveryService` delegates to `ResizeRecoveryService` or `SdxUpgradeRecoveryService`
* `ResizeRecoveryService` is not yet implemented
* `SdxUpgradeRecoveryService` is in flux, I'm not sure how far along it's implementation is

I think to properly interface this too, we'll need to define a `RecoveryResponse<SdxCluster>` or `RecoveryResponse<Datalake>`.

And I did some other renaming extraction to try to get things to line up.

Check out a preview of it [here](https://github.com/hortonworks/cloudbreak/pull/11709)

brycederriso replied:
Talked to Dan, we'll review the [second PR](https://github.com/hortonworks/cloudbreak/pull/11709) to make the interface change.

hreeve-cloudera commented on datalake/src/main/java/com/sequenceiq/datalake/service/recovery/RecoveryService.java:
```diff
@@ -0,0 +1,47 @@
+    private SdxUpgradeRecoveryService sdxUpgradeRecoveryService;
+
+    @Inject
+    private ResizeRecoveryService resizeRecoveryService;
```
nit: In the comment above you reference `SdxResizeRecoveryService`, but here it's `ResizeRecoveryService`.

brycederriso replied:
Thanks, fixed.

hreeve-cloudera commented on datalake/src/main/java/com/sequenceiq/datalake/service/resize/recovery/ResizeRecoveryService.java:
```diff
@@ -0,0 +1,46 @@
+     * Determines if it's possible to run a <em>Resize</em> recovery.
+     * @return true if resize recovery is possible.
+     */
+    public boolean canRecover() {
```
nit: I like the Javadocs you're adding. But this method is missing from the list of public methods.

brycederriso replied:
Thanks, I think the list of public methods is actually bad, they should be annotating each method individually.

I've tuned up the javadocs.

lnardai commented:
Hey guys! We've been keeping an eye on this PR for a couple of days now. It only got green today. 
@sodre90 was tasked with the review. Unfortunately he hadn't had the time to review it so far, but in such cases I would advise to wait at least a day or if it's urgent ping the reviewer about it honor the code Ownership and avoid unnecessary reverts.

lnardai commented:
Peter will do a postmerge review and update you on the findings here.

sodre90 commented:
Reverted. Please reopen and I will review early next week.

sodre90 commented:
Reverted. Please reopen and I will review early next week.

sodre90 commented:
First of all, please do the TODOs, then we can check this PR if it is finished.

sodre90 commented:
Also please use GH Draft feature until the PR is not ready.

brycederriso commented:
@sodre90 I've removed one of the todos as it's a nice to have, but two of the other ones are intended to be implemented in follow up issues for this one. 
I've added links to the follow up Jira in the comment. The idea is that this is going to be a big PR if it's all in one go and I wanted to make smaller PRs so they can be understood and reviewed more quickly.

Would it be better to make a large PR fixing the whole thing, or should I just remove the todos?
